src/Agent.Listener/Configuration/AutoLogonConfigurationManager.cs

```diff
@@ -0,0 +1,207 @@
+                Trace.Info("LogonAccount after transforming: {0}, user: {1}, domain: {2}", logonAccount, userName, domainName);
+
+                logonPassword = command.GetWindowsLogonPassword(logonAccount);
+                if (_windowsServiceHelper.IsValidCredential(domainName, userName, logonPassword))
+                {
+                    Trace.Info("Credential validation succeeded");
+                    break;
+                }
+                
+                if (command.Unattended)
+                {
+                    throw new SecurityException(StringUtil.Loc("InvalidWindowsCredential"));
+                }
+                    
+                Trace.Error("Invalid credential entered");
+                _terminal.WriteLine(StringUtil.Loc("InvalidWindowsCredential"));
+            }
+
+            bool isCurrentUserSameAsAutoLogonUser = _windowsServiceHelper.HasActiveSession(domainName, userName);            
+            if (!isCurrentUserSameAsAutoLogonUser)
+            {
+                if (!_autoLogonRegManager.DoesRegistryExistForUser(domainName, userName))
+                {
+                    Trace.Error($"The autologon user '{logonAccount}' doesnt have a user profile on the machine. Please login once with the expected autologon user and reconfigure the agent again");
+                    throw new InvalidOperationException(StringUtil.Loc("NoUserProfile", logonAccount));
+                }
+            }
+            _autoLogonRegManager.LogWarnings(domainName, userName);
+            
+            _autoLogonRegManager.UpdateRegistrySettings(domainName, userName);
+            _windowsServiceHelper.SetAutoLogonPassword(logonPassword);    
+            await ConfigurePowerOptions();
+            SaveAutoLogonSettings(domainName, userName);
+
+            if (!_windowsServiceHelper.HasActiveSession(domainName, userName))
+            {
+                RestartBasedOnUserInput(command);
+            }
+        }
+
+        private void RestartBasedOnUserInput(CommandSettings command)
+        {
+            Trace.Info("AutoLogon is configured for a different user than the current user. Machine needs a restart.");            
+            _terminal.WriteLine(StringUtil.Loc("RestartMessage"));
+
+            var shallRestart = command.GetRestartNow();
+            if (shallRestart)
+            {
+                var whichUtil = HostContext.GetService<IWhichUtil>();
+                var shutdownExePath = whichUtil.Which("shutdown.exe");
+
+                Trace.Info("Restarting the machine in 5 seconds");
+                _terminal.WriteLine(StringUtil.Loc("RestartIn5SecMessage"));
+                string msg = StringUtil.Loc("ShutdownMessage");
+                //we are not using ProcessInvoker here as today it is not designed for 'fire and forget' pattern
+                //ExecuteAsync API of ProcessInvoker waits for the process to exit
+                Process.Start(shutdownExePath, $"-r -t 5 -c {msg}");
```

Comment:
minor comment: note this flow becomes unit testable if you use process invoker but don't await